db/hints/sync_point.cc

```diff
@@ -0,0 +1,144 @@
+/*
+ * Copyright (C) 2021-present ScyllaDB
+ */
+
+/*
+ * This file is part of Scylla.
+ *
+ * Scylla is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Scylla is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Scylla.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <exception>
+#include <unordered_set>
+
+#include <seastar/core/simple-stream.hh>
+
+#include "db/hints/sync_point.hh"
+#include "gms/inet_address_serializer.hh"
+#include "idl/uuid.dist.hh"
+#include "idl/uuid.dist.impl.hh"
+#include "idl/replay_position.dist.hh"
+#include "idl/replay_position.dist.impl.hh"
+#include "idl/hinted_handoff.idl.hh"
+#include "idl/hinted_handoff.dist.hh"
+#include "idl/hinted_handoff.dist.impl.hh"
+#include "serializer.hh"
+#include "serializer_impl.hh"
+#include "bytes.hh"
+
+namespace db {
+namespace hints {
+
+// TODO: Encode host ID so that two sync points won't get mixed up
+
+static std::vector<sync_point::shard_rps> decode_one_type_v1(uint16_t shard_count, const per_manager_sync_point_v1& v1) {
+    std::vector<sync_point::shard_rps> ret;
+
+    if (size_t(shard_count) * v1.addresses.size() != v1.flattened_rps.size()) {
+        throw std::runtime_error("Invalid hint sync point");
+    }
+
+    ret.resize(shard_count);
+
+    auto rps_it = v1.flattened_rps.begin();
+    for (const auto addr : v1.addresses) {
+        for (uint16_t shard = 0; shard < shard_count; shard++) {
+            ret[shard].emplace(addr, *rps_it++);
+        }
+    }
+
+    return ret;
+}
+
+sync_point sync_point::decode(sstring_view s) {
+    // Decode from hex
+    bytes raw = from_hex(s);
+    if (raw.empty()) {
+        throw std::runtime_error("Invalid hint sync point");
```

Comment:
Let's mention that it's invalid because it's empty to make future debugging easier

Reply:
Improved the error message